lib/ts/authUtils.ts

```diff
@@ -0,0 +1,613 @@
+) {
+    if (session === undefined) {
+        throw new Error("This should never happen: getValidSecondaryFactors called without a session");
+    }
+    if (sessionUser === undefined) {
+        throw new Error("This should never happen: getValidSecondaryFactors called without a sessionUser");
+    }
+    logDebugMessage(`getValidSecondaryFactors called for ${factorIds.join(", ")}`);
+
+    const mfaInstance = MultiFactorAuthRecipe.getInstance();
+    if (mfaInstance !== undefined) {
+        if (createsNewLinkForTheSessionUser) {
+            // If we are linking the input user to the session user, then we need to check if MFA allows it
+            // Note that this is mostly but not exactly the same as a factor setup:
+            // we consider passwordless factors as already set up if the session user has a login method connected to them
+            logDebugMessage(`getValidSecondaryFactors checking if linking is allowed by the mfa recipe`);
+            return mfaInstance.checkIfLinkingAllowed(session, factorIds, userContext);
+        } else {
+            // If signing in will not change the user (no linking), then we can let the sign-in/up happen (if allowed by account linking checks)
+            logDebugMessage(`getValidSecondaryFactors allowing all factors because it'll not create new link`);
+            return factorIds;
+        }
+    } else {
+        logDebugMessage(`getValidSecondaryFactors allowing all factors because MFA is not enabled`);
+        // If MFA is not enabled, we allow the user to connect any secondary account to the session user.
+        return factorIds;
+    }
+}
+
+async function getValidFirstFactors(
+    factorIds: string[],
+    tenantId: string,
+    hasSession: boolean,
+    userContext: UserContext
+): Promise<{ status: "OK"; validFactorIds: string[] } | { status: "NOT_LINKING_NON_FIRST_FACTOR" }> {
+    let validFactorIds = [];
+    for (const id of factorIds) {
+        // This util takes the tenant config into account (if it exists), then the MFA (static) config if it was initialized and set.
+        let validRes = await isValidFirstFactor(tenantId, id, userContext);
+
+        if (validRes.status === "TENANT_NOT_FOUND_ERROR") {
+            if (hasSession) {
+                throw new SessionError({
+                    type: SessionError.UNAUTHORISED,
+                    message: "Tenant not found",
+                });
+            } else {
+                throw new Error("Tenant not found error.");
+            }
+        } else if (validRes.status === "OK") {
+            validFactorIds.push(id);
+        }
+    }
+
+    if (validFactorIds.length === 0) {
+        if (!hasSession) {
+            throw new SessionError({
+                type: SessionError.UNAUTHORISED,
+                message: "A valid session is required to authenticate with secondary factors",
+            });
+        } else {
+            return { status: "NOT_LINKING_NON_FIRST_FACTOR" }; // TODO: better name?
```

Comment:
We can't throw unauthorised because if that'd make the FE go into a retry loop since it does have a valid session. We need to either throw a normal error (500) or show a support code.
What this status means is:
- There is an active session (otherwise we throw unauthorised)
- For some reason this is still considered a first factor auth (e.g.: the account linking is off)
- This login method is not a valid first factor